Make timezone table independent from ORC

vuule opened the pull request:
## Description
Modifies the creation of timezone transition tables to make it usable outside of the ORC context:

- Remove the ORC epoch offset from the transition table.
- Use `timestamp_s` and `duration_s` instead of integral values in the table.
- Return table as `cudf::table` instead of the special `timezone_table` struct.
- Move the implementation/header out of `cudf::io::orc`.
- Split the header into C++ and CUDA parts. C++ API has public and `detail` counterparts.

Other:
Adapt ORC reader to changes to the timezone transition table.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

ttnghia commented:
I see that the timezone stuff is in the `cudf/orc/` folder? I believe that they are very independent, so please consider refactoring it out into somewhere else not in `cudf/orc/`, like `cudf/utilities/`. This would be useful for implementing other timezone-aware operations.

vuule commented:
> I see that the timezone stuff is in the `cudf/orc/` folder? I believe that it is very independent from ORC file format, so please consider refactoring it out into somewhere else not in `cudf/orc/`, like `cudf/utilities/`. This would be useful for implementing other timezone-aware operations.

This PR was a first draft just to unblock @shwina on the Python side. I've since split the headers to separate C++ and CUDA code and moved them to /include/detail. The namespace was also changed to cudf::detail.

vuule commented:
CC @shwina to confirm that new API and header location are suitable for Cython use.

shwina commented:
@vuule Thanks! while technically the new location is OK for Cython, can the function be moved out of the `detail` namespace?

vuule commented:
> @vuule Thanks! while technically the new location is OK for Cython, can the function be moved out of the `detail` namespace?

I placed it under `detail` because it's not really intended for end users. This is based on an offline discussion with @galipremsagar . I'm fine with moving it out of detail, the `detail` namespace is only meant to discourage direct use here.

shwina commented:
I think any code that is used outside of `libcudf` should be in its public API (even if that is just cuDF Python).

This avoids situations in the future where someone might decide to drop this code from `libcudf` because it's no longer being used internally, OR, a breaking API change is made but not considered breaking because it's not part of the public API. 

Perhaps it could be excluded from the doxygen-generated docs, so that libcudf consumers do not discover it?

ajschmidt8 reviewed (approved):
Approving `ops-codeowner` file changes

shwina commented on cpp/include/cudf/timezone.hpp:
```diff
@@ -0,0 +1,49 @@
+ */
+std::unique_ptr<table> make_timezone_transition_table(std::optional<std::string_view> tzif_dir,
+                                                      std::string_view timezone_name,
+                                                      rmm::cuda_stream_view stream);
```
Should this argument have a default value of `stream_default`?

shwina replied:
Actually, I think this argument should just be eliminated from the public API

mythrocks replied:
I've lost track of where we stand on eventually exposing `stream` parameters from non-`detail` public functions. Would removing the parameter mean having to add it back eventually?
Perhaps adding a default value is an acceptable compromise?

shwina replied:
cc: @vyasr

vuule replied:
yeah, now that I look at this as a regular libcudf API, it should not have a stream parameter. I'll add a `detail` counterpart that has the defaulted stream parameter.

vyasr replied:
Yeah I would remove the stream. As for where we stand, the answer is eventually but with no concrete timeline yet. We're making incremental progress towards it but for now I think the only public APIs exposing streams are ones that a privileged user (typically cuspatial) has explicitly requested as necessary.

mythrocks commented on cpp/include/cudf/detail/timezone.cuh:
```diff
@@ -0,0 +1,74 @@
+    } else {
+      auto project_to_cycle = [](timestamp_s ts) {
+        static constexpr duration_s cycle_s =
+          cuda::std::chrono::duration_cast<duration_s>(duration_D{365 * cycle_years + (100 - 3)});
```
ELI5: What's the calculation here? I'm afraid I don't follow.

vuule replied:
It codifies the leap year rules:


In our Gregorian calendar three criteria must be taken into account to identify leap years:
The year must be evenly divisible by 4;
If the year can also be evenly divided by 100, it is not a leap year;
unless...
The year is also evenly divisible by 400. Then it is a leap year.

I believe this translates to 100 (every fourth year) - 4 (every hundredth year ) + 1 (every four hundredth year) leap days in 400 years.

bdice replied:
Is this a rule we can get (or reference) from some other source? If we must implement it, is this codified in a way that is specific to ORC conventions or general to all time zone libraries?

To the greatest extent possible, cuDF should avoid implementing or defining its own rules for time math. We should always use (or reference) `libcudacxx`, other timezone libraries, an RFC, or similar.

vuule replied:
This is the definition of the Gregorian calendar. It's unrelated to ORC or timezones.
Given how specific this is (only multiples of 400 years have a fixed number of days, otherwise you need to know the year range) I doubt any library can compute this without computing number of days for each year individually (1-400 in this case, I think).

I can add a link to Gregorian calendar and clarification comments.

bdice replied:
I think I'm asking something different. I want a reference that states why we "search the 400-year cycle if outside of the file entries range" (in the comment below). We didn't come up with that, it's defined by some spec outside of libcudf.

vuule replied:
Ah, that. I actually don't know where this comes from. I'll try to find it and incorporate in some way.

vuule replied:
So, I've been digging around the interwebs and could not find anything on the use of 400 year cycle to compute time zone offsets in the future. 
The only thing I found is the name for this cycle - [Solar cycle](https://en.wikipedia.org/wiki/Solar_cycle_(calendar)). And even that is a seemingly overloaded term.
I added this to the names for the relevant constants, will incorporate it here as well.

vuule replied:
tried to make this clearer. Please let me know if it's still lacking

mythrocks commented on cpp/src/io/orc/stripe_enc.cu:
```diff
@@ -49,7 +49,7 @@ constexpr int scratch_buffer_size = 512 * 4;
 constexpr bool zero_pll_war = true;
 
-static __device__ __constant__ int64_t kORCTimeToUTC =
+static __device__ __constant__ int64_t orc_utc_epoch =
```
This constant looks similar to the `duration_s` in `stripe_data.cu`. I'm not sure I've understood why we have two constants for the same value.

vuule replied:
it is the same constant, just for the reader and the writer. I'll try to move it to a common header.

bdice commented on cpp/include/cudf/detail/timezone.cuh:
```diff
@@ -0,0 +1,74 @@
+{
+  if (tz_table.num_rows() == 0) { return duration_s{0}; }
+
+  cudf::device_span<timestamp_s const> ttimes(tz_table.column(0).head<timestamp_s>(),
```
Can we give this a longer name? Does `ttimes` mean `transition_times`? Let's try to increase clarity as much as possible in names.

bdice commented on cpp/include/cudf/timezone.hpp:
```diff
@@ -0,0 +1,49 @@
+ * Uses system's TZif files. Assumes little-endian platform when parsing these files.
+ *
+ * @param tzif_dir The directory where the TZif files are located
+ * @param timezone_name standard timezone name (for example, "US/Pacific")
```
Let's not use an obsolete time zone for our example. `US/Pacific` has been replaced by `America/Los_Angeles`.

http://www.timezoneconverter.com/cgi-bin/zoneinfo.tzc?s=default&tz=US/Pacific

bdice reviewed (changes requested):
Suggestions on time zone rules and naming. (Haven't dug much into the actual C++)

vuule commented:
Looks like I managed to straight-up murder this branch. Will open a new PR when I find the changes.

vuule commented on cpp/include/cudf/detail/timezone.cuh:
```diff
@@ -0,0 +1,74 @@
+      return first_larger - 1;
+    };
+
+    auto const file_entry_end = ttimes.begin() + (ttimes.size() - cycle_entry_cnt);
```
Made the name more specific, but kept the `snake_case` style. I hope it's at least a bit easier to see that it's a constant, based on the name.

vuule commented on cpp/include/cudf/detail/timezone.cuh:
```diff
@@ -0,0 +1,74 @@
+namespace cudf::detail {
+
+/**
+ * @brief Returns the GMT offset for a given date and given timezone table.
```
My understanding is that UT, UTC and GMT are equivalent for `second` precision. I have changed the TZif related code to use UT, but since ORC references UTC, I kept it there as such.

shwina reviewed (approved):
LGTM from Cython side!

bdice commented on cpp/include/cudf/timezone.hpp:
```diff
@@ -0,0 +1,49 @@
+// Cycle in which the time offsets repeat in Gregorian calendar
+static constexpr int32_t solar_cycle_years = 400;
+// Number of future entries in the timezone transition table:
+// Two entries per year, over the length of the Gregorian calendar's solar cycle
```
How do we know there are two entries per year? Is it guaranteed by the spec that timezones cannot have zero, one, three, or four entries?

vuule replied:
There are two entries because of Daylight Saving Time. If the time zone does not have DST, all future entries will have the same time offset. We could optimize this case in the future.

bdice replied:
It seems that we’re hardcoding an assumption that daylight savings time exists and results in exactly two changes per year. I don’t think either of those are safe assumptions.

vuule replied:
It is a safe assumption, since that's how `make_timezone_transition_table` works (currently). It will always include two entires per year in cycle. The case when DST does not exist, it will create 2 X cycle_years of the same offset, which is still correct, just slower than necessary.
We could separate the output columns into tzif entires and future entries. With that, we can have any number of elements in the future column. I would like to keep optimizations like this to a future PR. Let me know what you think.

bdice replied:
Document this ("it always includes two entries per year, assuming two transitions per year due to Daylight Savings Time") and any potential future work in a comment. That would be enough to resolve this and move forward.

vuule replied:
Agree with the suggestion, this on not clear from the API.
Added more detailed description of the transition table.

bdice reviewed (commented):
Code looks good. One minor question about the far, far future.

bdice commented on cpp/include/cudf/timezone.hpp:
```diff
@@ -0,0 +1,54 @@
+ *
+ * Uses system's TZif files. Assumes little-endian platform when parsing these files.
+ * The transition table starts with the entries from the TZif file. For timestamps after the file's
+ * last transition, the table includes entires that form a `solar_cycle_years`-year cycle (future
```
```suggestion
 * last transition, the table includes entries that form a `solar_cycle_years`-year cycle (future
```

bdice reviewed (approved):
Typo fix. Otherwise LGTM!

vuule commented:
/merge
